cpp/src/join/joining.cu

```diff
@@ -399,83 +393,121 @@ gdf_error construct_join_output_df(
+  int num_merging_join_cols = (merging_join_cols.num_columns() > 0 ? merging_join_cols.get_column(0)->size: 0);
+
+  std::vector<int> left_join_cols (num_cols_to_join);
+  std::vector<int> right_join_cols (num_cols_to_join);
+  std::vector<int> l_common_name_join_ind (num_merging_join_cols);
+  std::vector<int> r_common_name_join_ind (num_merging_join_cols);
+
+  if (num_cols_to_join > 0)
+  {
+      CUDA_TRY (cudaMemcpy((void *)left_join_cols.data(), (void *)join_cols.get_column(0)->data, 
+                           cudf::size_of(join_cols.get_column(0)->dtype)*num_cols_to_join, cudaMemcpyDeviceToHost));
+      CUDA_TRY (cudaMemcpy((void *)right_join_cols.data(), (void *)join_cols.get_column(1)->data, 
+                           cudf::size_of(join_cols.get_column(1)->dtype)*num_cols_to_join, cudaMemcpyDeviceToHost));
   }
-  for(int column_index = 0; column_index  < num_right_cols; column_index++){
-    GDF_REQUIRE(right_cols[column_index]->dtype != GDF_invalid,GDF_UNSUPPORTED_DTYPE);
+  if (num_merging_join_cols > 0)
+  {
+      CUDA_TRY (cudaMemcpy(l_common_name_join_ind.data(), merging_join_cols.get_column(0)->data, 
+                           cudf::size_of(merging_join_cols.get_column(0)->dtype)*num_merging_join_cols , cudaMemcpyDeviceToHost));
+      CUDA_TRY (cudaMemcpy(r_common_name_join_ind.data(), merging_join_cols.get_column(1)->data, 
+                           cudf::size_of(merging_join_cols.get_column(1)->dtype)*num_merging_join_cols , cudaMemcpyDeviceToHost));
   }
 
-  // Determine if requested output is the indices of matching rows, the fully
-  // constructed output dataframe, or both
-  bool const construct_output_dataframe{nullptr != result_cols};
-  bool const return_output_indices{(nullptr != left_indices) and
-                                   (nullptr != right_indices)};
+  std::vector <gdf_column*> tmp_left_cols;
+  std::vector <gdf_column*> tmp_right_cols;
 
-  GDF_REQUIRE(construct_output_dataframe or return_output_indices,
-              GDF_INVALID_API_CALL);
+  for (int i=0; i < num_left_cols; ++i)
+  {
+      tmp_left_cols.push_back(const_cast<gdf_column *> (left_cols.get_column(i)));
+  }
+
+  for (int i=0; i < num_right_cols; ++i)
+  {
+      if (std::find(r_common_name_join_ind.begin(), r_common_name_join_ind.end(), i)
+            == r_common_name_join_ind.end()) {
+          tmp_right_cols.push_back(const_cast<gdf_column *> (right_cols.get_column(i)));
+      }
+  }
 
-  auto const left_col_size = left_cols[0]->size;
-  auto const right_col_size = right_cols[0]->size;
+  cudf::table tmp_left_table = (tmp_left_cols.size()>0)? cudf::table (tmp_left_cols) : cudf::table{};
+  cudf::table tmp_right_table = (tmp_right_cols.size()>0)? cudf::table (tmp_right_cols) : cudf::table{};
+
+  CUDF_EXPECTS (0 != num_left_cols, "Left table is empty");
+  CUDF_EXPECTS (0 != num_right_cols, "Right table is empty");
+  CUDF_EXPECTS (nullptr != join_context, "Join context is invalid");
+
+  if (0 == num_cols_to_join)
```

Comment:
Move this up to where you first compute num_cols_to_join. If it's zero, skip all those loops above and just return.